weblate/api/views.py

```diff
@@ -2053,17 +2053,31 @@ def translations(self, request: Request, *args, **kwargs):
     def comments(self, request, *args, **kwargs):
         """Add a new comment to a unit."""
         unit = self.get_object()
+        user = request.user
 
-        if not request.user.has_perm("comment.add", unit.translation):
+        if not user.has_perm("comment.add", unit.translation):
             self.permission_denied(request)
 
         serializer = CommentSerializer(data=request.data, context={"unit": unit})
         serializer.is_valid(raise_exception=True)
 
         text = serializer.validated_data["comment"]
         scope = serializer.validated_data["scope"]
-        Comment.objects.add(request, unit, text, scope)
-        return Response(data={"result": True}, status=HTTP_201_CREATED)
+        timestamp = serializer.validated_data.get("timestamp")
+        user_email = serializer.validated_data.get("user_email")
+
+        if (timestamp is not None or user_email is not None) and (
+            not user.has_perm("project.edit", unit.translation.component.project)
+        ):
+            self.permission_denied(request)
+
+        if user_email is not None:
+            override_user = User.objects.filter(email=user_email).first()
+            if override_user is not None:
+                user = override_user
+
+        comment = Comment.objects.add(request, unit, text, scope, user, timestamp)
+        return Response(data={"id": comment.id}, status=HTTP_201_CREATED)
```

Comment:
I think having a proper model serializer and returning newly created serialized comment would be better.